### PR TITLE
Document how content generation works

### PR DESCRIPTION
**Pull request by jasonyavorsky**
This documents how content generation is implemented for embedded handbook content.

### CODE REVIEW

File: content/handbook/editing/content-generation.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Code
+
+The flow for content generation starts in [/src/lib/markdownToHtml.ts](https://github.com/sourcegraph/handbook/blob/main/src/lib/markdownToHtml.ts) in the function `insertGeneratedMarkdown`, which handles detecting a `{{generator:*}}` directive and passing it on to the appropriate markdown generation function.
+
+The markdown generation functions are in [/src/lib/generatedMarkdown.js](https://github.com/sourcegraph/handbook/blob/main/src/lib/generatedMarkdown.js). These functions are responsible for processing the data files and returning the markdown that will be inserted on the page. Typically these functions are iterating over objects, looking up values for other objects, in order to generate some sort of list that can be returned for embedding on the page.
+
+Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not.
```
Comment (jasonyavorsky) [inline content/handbook/editing/content-generation.md]:
```suggestion
Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not. If you add new data relationships, you should also add a check here.
```

**Comment by malomarrec** [comment]
Thank you, this is exactly what I was looking for.

File: content/handbook/editing/content-generation.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+```
+---
+data_source: [/data/code_hosts.yml, code hosts]
+data_source_2: [/data/product_teams.yml, product teams]
+---
+```
+
+The page supports up to two data sources (the second is optional).
+
+This will show, underneath "Edit this page" on the right hand side, "Edit team data". This is implemented in [/src/components/EditSection.tsx](https://github.com/sourcegraph/handbook/blob/main/src/components/EditSection.tsx).
```
Comment (jasonyavorsky) [inline content/handbook/editing/content-generation.md]:
```suggestion
This particular example will show, underneath "Edit this page" on the right hand side, "Edit code hosts data" and "Edit product teams data" (the first entry is the path, the second is what to display visually). This is implemented in [/src/components/EditSection.tsx](https://github.com/sourcegraph/handbook/blob/main/src/components/EditSection.tsx).
```

File: content/handbook/editing/index.md [hunk 3]
```diff
@@ -41,6 +41,10 @@ We don't expect everyone on the team to figure this out on their own. Other team
 - [Deleting a Page](delete-a-page.md)
 - [Editing the Handbook Product](editing-the-handbook-product.md)
 
+## Advanced topics
+
+- [Content generation](content-generation.md)
```
Comment (jasonyavorsky) [inline content/handbook/editing/index.md]:
```suggestion
- [How embedded content generation works](content-generation.md)
```

File: content/handbook/editing/content-generation.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  em: yink_teo
+```
+
+There are schema files for each of the data files at [/schema](https://github.com/sourcegraph/handbook/tree/main/schema) that set formats and whether different fields are required or not.
+
+When implementing new fields/structures, avoid creating new string values for things that you can look up via id or else we will have multiple sources of truth for the same values.
+
+## Code
+
+The flow for content generation starts in [/src/lib/markdownToHtml.ts](https://github.com/sourcegraph/handbook/blob/main/src/lib/markdownToHtml.ts) in the function `insertGeneratedMarkdown`, which handles detecting a `{{generator:*}}` directive and passing it on to the appropriate markdown generation function.
```
Comment (jasonyavorsky) [inline content/handbook/editing/content-generation.md]:
```suggestion
The flow for content generation starts in [/src/lib/markdownToHtml.ts](https://github.com/sourcegraph/handbook/blob/main/src/lib/markdownToHtml.ts) in the function `insertGeneratedMarkdown`, which handles detecting a `{{generator:*}}` directive and passing it on to the appropriate markdown generation function. If you add a new kind of generator you need to add the mapping here.
```

**Comment by jasonyavorsky** [comment]
@benvenker thought you might be interested, here's a real world example of the kind of content that could end up being an onboarding notebook.

**Comment by benvenker** [comment]
This is great, thanks @jyavorska

File: content/handbook/editing/content-generation.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+The markdown generation functions are in [/src/lib/generatedMarkdown.js](https://github.com/sourcegraph/handbook/blob/main/src/lib/generatedMarkdown.js). These functions are responsible for processing the data files and returning the markdown that will be inserted on the page. Typically these functions are iterating over objects, looking up values for other objects, in order to generate some sort of list that can be returned for embedding on the page.
+
+Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not. If you add new data relationships, you should also add a check here.
```
Comment (coury-clark) [inline content/handbook/editing/content-generation.md:29]: What is the philosophy if validations fail? Do we skip broken relationships? Do we throw an error on the entire generated block?

**Review by coury-clark — COMMENTED** [review]
Nice!

File: content/handbook/editing/content-generation.md (hunk 5 above)
Comment (jasonyavorsky) [reply content/handbook/editing/content-generation.md:29]: It returns the invalid object as JSON and fails the validation step in the build.

**Pull request merged**